### ISSUE TITLE
Tracking issue for `concat_idents`

### ISSUE THREAD

**Issue by aturon** [rust-lang/rust#29599]
Tracks stabilization for the `concat_idents` macro.

**Update(fmease)**: Please see https://github.com/rust-lang/rust/issues/124225#issue-2255069330 (an alternative to this feature).

**Comment by retep998** [comment]
`concat_idents` is so fundamentally useless at the moment because macros cannot be used in ident positions.

**Comment by Toby-S** [comment]
Yeah it is pretty much useless right now (see #12249 #13294).

**Comment by nrc** [comment]
Relevant blog for future plans: http://www.ncameron.org/blog/untitledconcat_idents-and-macros-in-ident-position/

**Comment by eddyb** [comment]
A solution that has been thrown around on IRC (but never ended up anywhere else AFAIK):
Have a macro invocation form for "early expansion", i.e. the following would work without macros in ident position:
```rust
macro_rules! wrap {
    ($name:ident) => { struct concat_idents!!(Wrapped, $name)($name); }
}
```
The "early expansion" `macro_name!!(args)` syntax was originally `$*macro_name!(args)` and is a subject of bikeshed.

If we move towards having all macros produce token streams that are parsed on-demand, such "early expansion" could be used in many more locations without adding macro support for each one.
Generic parameters, ADT fields, function signatures and match arms come to mind - there are so many recursive push-down hacks macros can end up with, just to construct lists of things and whatnot, that would simply not be necessary with "early expansion".

The only disadvantage is that `concat_idents!!` and friends would only work inside the RHS of `macro_rules!`, but I don't really see why you would need to use such a capability outside of a macro.

**Comment by skade** [comment]
https://github.com/mikkyang/rust-blas uses it for quick notation of FFI functions that follow a certain scheme: https://github.com/mikkyang/rust-blas/blob/master/src/prefix.rs https://github.com/mikkyang/rust-blas/blob/master/src/matrix/ll.rs

**Comment by skade** [comment]
Well, given that https://github.com/mikkyang/rust-blas/issues/12 is a nicer solution within the current language, I'd also like to raise my hand for "useless".

**Comment by drbo** [comment]
I think that generating FFI bindings is enough of a use that this should be stabilized.

**Comment by nrc** [comment]
I think we should never stabilise the current version - as others have noted it is useless.

**Comment by aidanhs** [comment]
Since this is the tracking issue, other who stumble across it might be interested in https://github.com/SkylerLipthay/interpolate_idents which works on nightly rust.

**Comment by retep998** [comment]
Unfortunately that solution will always depend on nightly Rust (until the day that syntax extensions become stable and pigs fly).

**Comment by skade** [comment]
Given that the consensus seems to be "useless" and theres nicer solutions (e.g. [1], maybe we can remove this feature?

[1] https://github.com/mikkyang/rust-blas/pull/12/files

**Comment by eddyb** [comment]
See rust-lang/rfcs#1628.

**Comment by aidanhs** [comment]
@retep998 yes, just noting it as a stopgap.

@skade it's the implementation rather than the idea that's useless. If `concat_idents` is removed rather than fixed, something else needs to fill its place (your solution doesn't work for all use-cases). I like the look of the rfc @eddyb linked, will follow that.

**Comment by skade** [comment]
@aidanhs that still means that we should drop the feature to make sure no one uses it.

I agree that my solution doesn't cover all edge-cases, but all uses that I've currently seen in the wild. Also, but this is outside of my competence to decide, I think this is a perfect case for the use of a code generator.

**Comment by aidanhs** [comment]
@skade afaict nobody *can* use it, so that's not a big concern...but equally there's no reason to keep it around. My main motivation for leaving it was as a reminder that people do want the functionality, but I'm content to follow the RFC and have changed my position to "I don't mind either way".

Regarding use cases, here's [the motivating example](https://github.com/aidanhs/ayzim/blob/713497c264/src/cashew.rs#L153-182) that brought me to this issue in the first place. Codegen's awesome for things like phf, but it feels a bit like overkill when I just want to generate some repetitive methods to extract tuples from enum variants. I guess it's down to personal preference.

**Comment by ExpHP** [comment]
An amusing fact about the current implementation of `concat_idents!` is that it will accept an empty argument list, making it possible to construct the empty string as an identifier:

```
error[E0425]: cannot find value `` in this scope
 --> src/main.rs:5:5
  |
5 |     concat_idents!();
  |     ^^^^^^^^^^^^^^^^^ not found in this scope
```

**Comment by petrochenkov** [comment]
>construct the empty string as an identifier

Empty string is used as a reserved identifier with special meaning in several places in the compiler, so this can potentially cause issues.

**Comment by jonhoo** [comment]
The linked RFC has been closed as postponed. Should `concat_idents!` now be removed? Is there any chance of seeing something in its place that supports generating identifiers?

**Comment by durka** [comment]
The RFC should be reopened. It never should have been closed, as no better solution was proposed.

**Comment by skade** [comment]
@durka I don't agree. There's ample reason to expect a new one when moving towards macros 2.0. RFCs are not necessarily closed to be replaced immediately, they are closed to stop following that line of discussion. Future RFCs might refer to it.

**Comment by alexreg** [comment]
I'm with @durka on this.

**Comment by eddyb** [comment]
@skade I don't think much has happened regarding macros 2.0 design except for maybe some implementation experience - e.g. @jseyfried might be aware of potential issues with early expansion.

**Comment by durka** [comment]
@skade, we are moving towards macros 2.0 rust-lang/rfcs#1584 was merged and `#![feature(decl_macro)]` exists on nightly. When is the appropriate time to reopen all the concerns that were tabled as "wait for macros 2.0"?

**Comment by skade** [comment]
@eddyb @durka that is both true, my point is that this probably merits a new RFC, based on this one, except of just reopening.

**Comment by dtolnay** [comment]
I implemented a stable concat_idents approach that works with any compiler >=1.15.0. https://github.com/dtolnay/mashup

**Comment by alexreg** [comment]
@dtolnay Black magic! Good job though. :-) Does Macros 1.2 support this yet? CC @jseyfried

**Comment by ExpHP** [comment]
> Does Macros 1.2 support this yet?

If Macros 1.2 added support for any form of early expansion, everybody would know about it.  I mean, it'd make international news.

**Comment by alexreg** [comment]
Hah. I know it was being discussed properly at one point... guess it never got too far!

**Comment by dtolnay** [comment]
I would like to propose stabilizing [`concat_idents!`](https://doc.rust-lang.org/std/macro.concat_idents.html) as it currently exists. As discussed at length in this thread, `concat_idents` does not solve all use cases for concatenated identifiers especially as a consequence of no macro expansion in identifier position. Nevertheless, this macro does a useful thing and [has its uses](https://github.com/search?l=Rust&q=%22feature+concat_idents%22&type=Code) for putting together certain types of convenient APIs.

Basically I don't feel the need to push for macro expansion in more places before stabilizing use of `concat_idents` in expression position.

For more elaborate use cases of concatenated identifiers I believe directing people to use [`mashup`](https://github.com/dtolnay/mashup) is acceptable.

@rfcbot fcp merge

**Comment by ExpHP** [comment]
Stabilization? This? *Really*?

Uhh, in that case, I guess I'll throw up a PR to fix the empty identifier bug (since I basically found it by looking at the source code).

**Comment by alexreg** [comment]
@dtolnay Sounds reasonable. Using it in identifier places can come when my PR for hygiene escaping lands. :-)

**Comment by durka** [comment]
Maybe with a big banner in the docs saying "this probably isn't what you
want!!!" and pointing to mashup.

**Comment by eddyb** [comment]
@joshtriplett What do you mean? This macro can't be implemented in a library, it's already built-in.
*However*, macro invocations *are not* allowed in identifier positions which simplifies many things (how do you parse a macro invocation when its name could be produced by another macro?), and this is where "early expansion" comes in, because then you're "just" expanding arbitrary tokens, *not* ASTs.

**Comment by durka** [comment]
@joshtriplett I don't think `concat_idents` is at fault, really. The issue is that macros can't be expanded in certain positions. Something would have to change around that, or around eager expansion, to be able to use `concat_idents` for something like constructing the names of getter/setter functions. The other issue is hygiene, but that is going to be fixed with proc macros in another way ([punted for now](https://internals.rust-lang.org/t/help-stabilize-a-subset-of-macros-2-0/7252)).

**Comment by alexreg** [comment]
@eddyb Ah, I thought it was just a hygiene problem... Looks like it's a hygiene *and* early-expansion problem.

**Comment by mark-i-m** [comment]
Perhaps change the name to reflect that it probably isn't what you want?

**Comment by ExpHP** [comment]
Is there a reason why it can't be used in patterns?

**Comment by ExpHP** [comment]
Oh and FYI you can construct keywords as identifiers, too =P

**Comment by cramertj** [comment]
@rfcbot concern naming

Practically every use I've had for `concat_idents` has required being able to use macros in identifier position, so if I discovered a stable macro called `concat_idents` I'd certainly expect this to work. Can we come up with a better name for the macro that is suggestive of its limited usability, while also leaving the door open to a macro that "does what you want" such as `mashup` being called `concat_idents!`?

**Comment by sfackler** [comment]
The macro does what it says on the tin - it concatenates idents. The changes required to make it work in positions it doesn't yet aren't in the macro, they're in the rest of the compiler.

**Comment by alexreg** [comment]
@cramertj `mashup` is a big hack, right now, as I see. Anyway, it's a 3rd-party crate, and I don't see why we should be concerned about official Rust identifiers clashing with 3rd-party ones. Once eager-expansion and hygiene land, it *will* do what we all want.

**Comment by ExpHP** [comment]
@sfackler if that was directed to me, pattern macros already exist.

If not, uh... carry on!

**Comment by durka** [comment]
I'm not confident eager expansion is coming, and not sure what "hygiene
landing" means.

**Comment by durka** [comment]
Are there actual uses that can't be replaced by mashup? Mashup is a hack
for now, but it should be fine once function-like proc macros land.

Unscientifically, from the first page of the search results @dtolnay
linked, 7 repos don't actually use the macro (presumably they added the
feature and then discovered it was useless).

Maybe concat_idents is tantalizing but useless and should be deprecated.

On Wed, May 2, 2018 at 10:46 PM, Michael Lamparski <notifications@github.com

**Comment by retep998** [comment]
I'd love to see someone do a demonstration where they adjust `winapi`'s macros to use `mashup` and eliminate the redundant inputs I have to specify due to the lack of a useful `concat_idents`.

**Comment by joshtriplett** [comment]
@eddyb I can understand not being able to produce a macro invocation with a macro, for instance. But why does that prevent, for instance, generating the name of a function or variable?

**Comment by dtolnay** [comment]
Sorry about the search results @durka, obviously :broken_heart: GitHub search. I clicked through the first 8 pages and found the following diverse and wonderful use cases. These crates are using concat_idents because it does exactly what is needed for their situation. I am happy to dig through all 27 pages if this does not change your mind about the macro being "tantalizing but useless and should be deprecated."

I believe `concat_idents!` is a valuable building block for macros, much like `stringify!`, despite not solving **everybody's** use cases **all the time**.

---

- [`JohnColanduoni/rust-vk-middle`](https://github.com/JohnColanduoni/rust-vk-middle/blob/355a41605903beadddfc5899b8f68b3da9617698/src/middle/src/icd/mod.rs#L185)

    ```rust
    stringify!($func_name) => mem::transmute::<concat_idents!(PFN_, $func_name), _>(Some(*&$func_name::<I>)),
    ```

- [`quininer/tox-rs`](https://github.com/quininer/tox-rs/blob/d1bd822fd386d2e3b22621acbabef8e20bada4f0/src/utils/macros.rs#L81)

    ```rust
    concat_idents!($prefix, _callback_, $event)($core, concat_idents!(on_, $event), $tx);
    ```

- [`s3bk/math_traits`](https://github.com/s3bk/math_traits/blob/3691e505aad347f5ea986b535b40fab5e43cceef/src/simd.rs#L14)

    ```rust
    transmute(concat_idents!($pre, $name, $post)( $( transmute($arg) ),* ))
    ```

- [`mikkyang/rust-lapack`](https://github.com/mikkyang/rust-lapack/blob/5af24369cc3d0f173d31afbb6e9695dfabf2b5ad/src/prefix.rs#L9)

    ```rust
    (Complex64, $f: ident) => (concat_idents!(z, $f));
    ```

- [`derekchiang/rust-hyperdex`](https://github.com/derekchiang/rust-hyperdex/blob/a4dd0478c5502763bb63b46a4d9d76164e16fdf0/src/client.rs#L997)

    ```rust
    concat_idents!(hyperdex_client_, $fn_name)(*inner_client.ptr,
    ```

- [`tasogare3710/ecma262_syntax`](https://github.com/tasogare3710/ecma262_syntax/blob/55e9dbe4ce2b45c7fdae123dcb5b5cf6e2767847/src/parser/internal/es/string_literal.rs#L218)

    ```rust
    concat_idents!(read_, $quote, _string_character)($ts, $string_chars_pos, $buff)?;
    ```

- [`teepee/teepee`](https://github.com/teepee/teepee/blob/63dff50897ae215f60a2cb4c668e7f83fd0da88e/src/headers/internals.rs#L881)

    ```rust
    (1 s $x:ident) => (Typed::Single(Box::new(concat_idents!(d1_s_, $x)())));
    ```

- [`pi-pi3/julia-rs`](https://github.com/pi-pi3/julia-rs/blob/01704dbdf741b4440112a5af998cb946cba08412/src/api/value.rs#L572)

    ```rust
    unsafe { Value::new_unchecked(concat_idents!(jl_box_, $t2)($fn)) }
    ```

- [`torchrs/torchrs`](https://github.com/torchrs/torchrs/blob/51a5857376825141cba2cf34260e49c291b72c43/src/storage.rs#L12)

    ```rust
    unsafe { concat_idents!($thname, _retain)(self.t)};
    ```

- [`jkarns275/f128`](https://github.com/jkarns275/f128/blob/90a882adfe9266cc01725762dce6316ad86f44d1/src/f128_derive.rs#L340)

    ```rust
    unsafe { f128::from_arr(concat_idents!($sm, _to_f128)(small)) }
    ```

- [`Limeth/euclider`](https://github.com/Limeth/euclider/blob/6c9b2615796c6a558baf830336898e68ad10a3c1/src/scene.rs#L1129)

    ```rust
    concat_idents!(blend_function_, $name)()
    ```

- [`dylanmckay/engine`](https://github.com/dylanmckay/engine/blob/47dae111e9863a4beb7edbd161f92f74517df606/src/gfx/src/gl/shader.rs#L218)

    ```rust
    let f = concat_idents!(Uniform1, $suffix);
    ```

- [`framlog/dpdk-rust`](https://github.com/framlog/dpdk-rust/blob/fcf9dd49ae7dc4c89faddcd260cb9ef6f3490035/src/lib.rs#L11)

    ```rust
    ($id: ident) => { unsafe { concat_idents!(per_lcore_, $id) } };
    ```

- [`eerimoq/rafiki`](https://github.com/eerimoq/rafiki/blob/99591be0c6a271970ab14e6ddafe56a99e6ec53f/src/lib.rs#L76)

    ```rust
    match concat_idents!($callback, _impl)(
    ```

**Comment by dtolnay** [comment]
Through page 13:

---

- [`Wilfred/remacs`](https://github.com/Wilfred/remacs/blob/1ac2bcf00b7511d1e7258e909b152a426af241d6/rust_src/src/lisp.rs#L1785)

    ```rust
    defsubr(concat_idents!(S, $f).as_ptr());
    ```

- [`michaelfairley/ezing`](https://github.com/michaelfairley/ezing/blob/1a259e8d9c41f75b271bed656fef8613f42c17a5/src/lib.rs#L260)

    ```rust
    #[test] fn in_() { test(concat_idents!($name, _in)); }
    ```

- [`Limeth/cubeb-rs`](https://github.com/Limeth/cubeb-rs/blob/a8ca0053c60ef17f4a76aced5d93c46eb92fdc2c/src/lib.rs#L64)

    ```rust
    let error_code = concat_idents!($ffi_prefix, $name)($field$($other_fields)*, result);
    ```

- [`redox-os/drivers`](https://github.com/redox-os/drivers/blob/513ddd9eb5f3d1b32da1e795f97894a0666c4d12/alxd/src/device/mod.rs#L141)

    ```rust
    ((($x) >> concat_idents!($name, _SHIFT)) & concat_idents!($name, _MASK))
    ```

- [`mattmacy/ruFAST`](https://github.com/mattmacy/ruFAST/blob/3c93b2fa81d0b2f5942c705534461cfac36f77d6/src/lib.rs#L28)

    ```rust
    unsafe { $name { ir : concat_idents!(FAST, $name, New)() } }
    ```

- [`jihoonson/arrow-rs`](https://github.com/jihoonson/arrow-rs/blob/d3527117f3f0e2f774654d9902b0ce2be3f5aacd/src/types/primitive.rs#L131)

    ```rust
    raw_builder: *mut concat_idents!(Raw, $builder_name)
    ```

- [`redox-os/kernel`](https://github.com/redox-os/kernel/blob/670d7b00d3d927e3c1583958e97c16a7db0c3851/src/arch/x86_64/paging/mod.rs#L143)

    ```rust
    virt_addr >= & concat_idents!(__, $n, _start) as *const u8 as usize &&
    ```

**Comment by dtolnay** [comment]
- [`bnavetta/rcmark`](https://github.com/bnavetta/rcmark/blob/ae92fb268933d5c0f4f98d39e58a0b4a78a08f3a/src/node.rs#L17)

    ```rust
    let raw_fn = concat_idents!(cmark_node_get_, $prop);
    ```

- [`zitsen/unqlite.rs`](https://github.com/zitsen/unqlite.rs/blob/3752ef5c89873d8fc073e3eff3be16055156a6aa/src/kv_cursor.rs#L188)

    ```rust
    concat_idents!(unqlite_kv_cursor_, $i)($($e),*)
    ```

- [`lkpdn/flota`](https://github.com/lkpdn/flota/blob/aea8a528ea654e9a95ecb36dbd579da0674ff1ae/src/virt/mod.rs#L28)

    ```rust
    pub unsafe fn raw(&self) -> concat_idents!($prefix, Ptr) {
    ```

- [`jonhoo/ordsearch`](https://github.com/jonhoo/ordsearch/blob/b17cb0b3f5089112145af25dfad6bbc8376c408a/src/lib.rs#L502)

    ```rust
    let mapper = concat_idents!(nodup_, $v);
    ```

- [`tz-lom/libarchive-sys`](https://github.com/tz-lom/libarchive-sys/blob/9b85cb4bb0b19df3acd2c9a61b74f359ee620f3e/src/lib.rs#L526)

    ```rust
    let sec = (concat_idents!(archive_entry_, $apiname))(self.entry);
    ```

**Comment by petrochenkov** [comment]
@rfcbot concern: Audit hygienic context of the produced identifier, add relevant tests

Context of the ident should probably be the context of `concat_idents!` macro invokation.

**Comment by alexreg** [comment]
@duroa Hygiene opt-out/escaping landing, I mean.

**Comment by durka** [comment]
OK, that's a lot more uses than I thought! I guess it's hard to replace
them all with mashup because you'd have to generate so many macros?

**Comment by alexreg** [comment]
@durka Yeah. I feel `concat_idents` + early expansion + hygiene escaping is the more elegant and general solution that fits all potential use cases.

**Comment by dtolnay** [comment]
@petrochenkov currently `concat_idents` cannot refer to a captured local variable, only an item and items do not have hygiene. Do you think `concat_idents` is different enough from other cases of non-atomic code fragments that we need to block stabilization until the non-atomic hygiene story is worked out? I would think that can happen later and does not need to be a blocker.

**Comment by petrochenkov** [comment]
@dtolnay 
>currently concat_idents cannot refer to a captured local variable

That's a bug.

>and items do not have hygiene.

All identifiers have hygiene with macros >= 1.2!
`concat_idents` looks like a typical use case for `Span::call_site` hygiene.

>Do you think concat_idents is different enough from other cases of non-atomic code fragments that we need to block stabilization until the non-atomic hygiene story is worked out?

`concat_idents` is not special, #50122 (or rather its minimal subset about call-site for macro invocations) is a blocker for it the same degree as for Macros 1.2 in general.

**Comment by durka** [comment]
What should happen with concat_idents!($a, $b) when $a and $b come from
different hygiene contexts?

On Thu, May 3, 2018 at 1:55 PM, Vadim Petrochenkov <notifications@github.com

**Comment by petrochenkov** [comment]
@durka 
I'd expect contexts of components to be completely lost during concatenation, it's a purely string-content based operation after all.

**Comment by durka** [comment]
OK, that makes sense. I just remember that being a question that came up
around this macro before.

On Thu, May 3, 2018 at 2:23 PM, Alexander Regueiro <notifications@github.com

**Comment by nrc** [comment]
@concern premature

I don't think we should stabilise concat_idents, between the macro system and the macro itself, this is not the macro that most macro authors want, but it is occupying prime naming real estate. I would prefer to wait until we have a macro that does do what people want before stabilising something.

Is there a strong reason to stabilise this now when we've been OK without stabilising it for years? It does seem like a really sub-optimal thing to stabilise.

**Comment by durka** [comment]
So the main issue with "it doesn't do what macro authors want" is that it
only works in expressions, right? It doesn't work for defining things
because macros aren't accepted in ident position. Some possible mitigations:

1. Just make that extremely clear in the documentation
2. Change the name somehow to reflect it (concat_idents_for_expr?
make_expr_name?). If eager expansion actually happens the name would be
awkward, but we could deprecate and change the name back in that case.
3. Make the error message better when discovering a macro invocation in
ident position (currently it's just unexpected `!` with no notes or helps)

**Comment by alexreg** [comment]
1. or 3. sound best right now.

**Comment by dtolnay** [comment]
I continue to believe this macro in its current form serves as a valuable building block for API design despite being limited to solving a specific set of use cases and not them all. There comes a time in an experienced macro author's life when they have seen many dozen situations that require exactly this functionality and wonder what the holdup could be after [almost 7](https://github.com/rust-lang/rust/commit/513276e595f8a8e988ec824a36b34cebe2aeaab4) years. But I can see that changing so many minds is not going to be a good use of two teams' time. Thanks for the discussion everyone! :slightly_smiling_face: 

@rfcbot fcp cancel

**Comment by devsnek** [comment]
has there been any movement on being able to use this in definitions?

**Comment by cramertj** [comment]
@devsnek nope, most use cases can be solved by using https://github.com/dtolnay/paste though.

**Comment by devsnek** [comment]
@cramertj will that allow `fn concat_idents!(x, y)() {}`? If so that's great, but it still feels like concat_idents not working on identifiers in definition positions is an abstraction leak of however the parser works.

**Comment by cramertj** [comment]
@devsnek `concat_idents` basically does not work, and there is no intention to stabilize it as it exists today.

**Comment by DzenanJupic** [comment]
I also had that problem and just wrote and published a crate that provides a working solution:
https://crates.io/crates/concat-idents
https://github.com/DzenanJupic/concat-idents

The syntax is a bit different from the std-solution:
```rust
concat_idents!(<IDENT> = <IDENT1>, <IDENT2>, <IDENT3>, [...] {
    // --snip--
});
```
But it works and can be used to, for example, generate tests:

```rust
macro_rules! generate_test {
   ($method:ident($lhs:ident, $rhs:ident)) => {
       concat_idents!(test_name = $method, _, $lhs, _, $rhs {
           #[test]
           fn test_name() {
               let _ = $lhs::default().$method($rhs::default());
           }
       });
   };
}

#[derive(Default)]
struct S(i32);

impl Add<i32> for S {
    type Output = S;

    fn add(self, rhs: i32) -> Self::Output {
        S(self.0 + rhs)
    }
}
impl Sub<i32> for S {
    type Output = S;

    fn sub(self, rhs: i32) -> Self::Output {
        S(self.0 - rhs)
    }
}

generate_test!(add(S, i32));
generate_test!(sub(S, i32));
```
In this case, the last two lines would expand to:
```rust
#[test]
fn add_S_i32() {
    let _ = S::default().add(i32::default());
}
#[test]
fn sub_S_i32() {
    let _ = S::default().sub(i32::default());
}
```

As long as the resulting identifier is valid, you can use a combination of identifiers, underscores, integers, and booleans.
Currently, it's not possible to define multiple identifiers in one macro call, but I'll add this feature soon.

I would love to hear your feedback :)

**Comment by sdroege** [comment]
> I also had that problem and just wrote and published a crate that provides a working solution:
> https://crates.io/crates/concat-idents
> https://github.com/DzenanJupic/concat-idents

That sounds very similar to the [paste](https://crates.io/crates/paste) crate

**Comment by mimoo** [comment]
Is there a way to use this but with a type `ty`? Perhaps by extracting the `ident` from `ty`? But I can't find a way to do that in a `macro_rules`.

**Comment by ExpHP** [comment]
Once a macro rules macro parses a token as anything other than `tt` or `ident` it becomes completely opaque.  If you need to do something more complicated like get the `Foo` from `Foo<A, B>` then you should consider writing a proc macro and using the `syn` and `quote` crates.

**Comment by joshtriplett** [comment]
If we were to keep/stabilize this, it would need to work in the fashion people expect it to work (in place of an identifier).

**Comment by reitermarkus** [comment]
I'd like for this to also support concatenating with integer literals, e.g.

```rust
const MY_CONST1: u32 = 1;
const MY_CONST2: u32 = 2;

concat_idents!(MY_CONST, 2)
```

**Comment by AlexisTM** [comment]
This is especially useful for embedded when we have many names we do not have control over. (register names) where most of the difference is a number.

**Comment by dvdsk** [comment]
It would be nice if we could use `concat_idents!` would optionally accept `self.` as first argument. This would allow accessing member variables and functions.

**Comment by reitermarkus** [comment]
@dvdsk, I think this is already covered by https://github.com/rust-lang/rust/issues/29599#issuecomment-1572068214. If the macro works in `ident` positions, I assume you can simply use e.g. `self.concat_idents!($field, 0)`.

**Comment by dvdsk** [comment]
> @dvdsk, I think this is already covered by [#29599 (comment)](https://github.com/rust-lang/rust/issues/29599#issuecomment-1572068214). If the macro works in `ident` positions, I assume you can simply use e.g. `self.concat_idents!($field, 0)`.

Ah it was my impression that `ident` positions referres to `fn concat_idents!(macro_args)(function args)`. I am not sure if that also implies `self.concat_idents!()`. That seemed more difficult to implement since `.` is not usually followed by a macro. It might even be confusing to some, is it calling a _member_ macro?

**Comment by tgross35** [comment]
Three comments:

1. What are the problems with stabilizing this as-is and later adding the expected features? This does not seem to 
2. Are there problems with special casing this macro to allow only its use in the identifier position? Even only within other macros, so that the relationship between `macro_rules!` and `concat_idents!` is about the same as `paste!` and paste's `[<...>]` syntax. Obviously this would not be trivial to implement...
5. This should be able to unstringify string literals, `concat_idents!(foo, "bar")` -> `foobar`, for macros where it is logical for a user to specify a string but an identifier needs to be created from it

**Comment by SylvKT** [comment]
From [the docs](https://doc.rust-lang.org/std/macro.concat_idents.html):
> Also, as a general rule, macros are only allowed in item, statement or expression position. That means while you may use this macro for referring to existing variables, functions or modules etc, you cannot define a new one with it.

Is there any particular rationale for disallowing identifier concatenation in the definition of items? I really hate using `paste`.

**Comment by SimonSapin** [comment]
It’s not that they’re excluded explicitly. It’s the other way around: parts of the language that expect an item, statement, or expression each accept various kinds of syntax, one of which being a macro invocation. In parts of the language that expect an identifier, the syntax is more restrictive (only accepting a plain or raw identifier token).

Accepting a macro invocation everywhere an identifier is accepted would be tricky at best because macro invocations themselves start with an identifier. `paste` adds dedicated `[<`…`>]` syntax specifically to avoid these ambiguities. *Maybe* accepting a macro invocation in a few more specific places like the name of an `fn` item is possible, I don’t know. I suspect even coming up with a concrete and plausible proposal would take some work.

**Comment by DemiMarie** [comment]
What about having `concat_idents` be a reserved word in Rust 2024?  That would eliminate the ambiguity, because `concat_idents!` would no longer be a macro.  Instead, it would be dedicated syntax that merely _appeared_ to be a macro.

Of course, this might cause parsing problems.  In particular, I doubt it would be compatible with an LALR(1) or LR(1) grammar.

**Comment by petrochenkov** [comment]
At this point it would probably be better to implement ident concatenation as a ["metavariable expression"](https://github.com/rust-lang/rust/issues/83527), like `${concat_idents(a, b, c)}` or something.

Then in a code produced by `macro_rules` and parsed by Rust parser it will already be represented as a single identifier, so it will work in any position.

**Comment by c410-f3r** [comment]
I think it would be nice to have something like `${concat_idents(a, ${concat_idents(b, c)})`, i.e., a mechanism to eagerly evaluate a metavariable expression.

**Comment by tgross35** [comment]
I'm sure there is a good answer but I don't see it here - why do we need `concat_idents` as a standalone macro in the first place? In theory it should be possible to have macro bodies have a syntax that concatenates things, like `foo_##bar` in C macros or `[<foo_ $bar>]` in `paste`. I guess this is in line with the metavar idea.

Also, it is extremely helpful to be able to convert between upper, lower, snake, and camel cases identifiers in paste. I sure wouldn't mind having this in std somehow if there were a good syntax...

**Comment by safinaskar** [comment]
🚀 I found a way to evaluate `concat_idents` (and few other built-in macros) **before** evaluating other macro, which takes `concat_idents` as an argument! I. e. I found a way to evaluate `a!(concat_idents!(...))` such way, that `concat_idents` evaluates before `a`. Answer is crate https://crates.io/crates/with_builtin_macros !!!

**Comment by tgross35** [comment]
This feature will be removed in the near future (https://github.com/rust-lang/rust/pull/142704). The (unstable) `macro_metavar_expr_concat` feature provides a more usable version of this functionality, tracked at https://github.com/rust-lang/rust/issues/124225.
